meta-balena-common/recipes-core/initrdscripts/files/migrate

```diff
@@ -26,32 +26,19 @@
 # shellcheck disable=SC1091
 . /usr/sbin/balena-config-defaults
 
-memory_check() {
-    _size="${1}"
-    sync && echo 3 > /proc/sys/vm/drop_caches
-    swapoff -a && swapon -a
-    _memfree=$(awk '/MemFree/{free=$2} END{print (free*1024)}' /proc/meminfo)
-    if [ "${_memfree}" -lt "${_size}" ]; then
-        relocate_log_mount
-        fail "Not enough memory: requested ${_size}, available ${_memfree}"
-    fi
-    return 0
-}
+# override fail() from os-helpers-logging to sync logs to disk and poweroff
+# instead of exiting, which would result in a panic and may prevent logs from
+# being written
+fail() {
+    error "$1"
 
-relocate_log_mount() {
-    # Preserve log file in rootfs
-    # Move the /run mountpoint to the rootfs mountpoint
-    if [ -d "${ROOTFS_DIR}/run" ]; then
-        if mountpoint '/run' > /dev/null; then
-            mount -n -o move /run "${ROOTFS_DIR}/run"
-        fi
-    else
-        # Older releases do not create /run so use /tmp instead
-        if mountpoint '/run' > /dev/null; then
-            mount -n -o move /run "${ROOTFS_DIR}/tmp"
-        fi
-    fi
-    sync "${ROOTFS_DIR}/run" "${ROOTFS_DIR}/tmp" || true
+    # ensure the boot partition is safely unmounted to prevent corruption,
+    # which can cause problems if a user retries the migration
+    disable_persistent_logging
+    umount -a
+    sync
+
+    poweroff -f && sleep infinity
```

Comment:
A poweroff will leave a remote device unreachable - the device needs a reboot. Returning an error exit from an initramfs module will already have that effect, I'd argue we don't need more than sync & umount here.

Also, instead of overriding the `fail` script, why not use an exit trap that will sync & umount?

Reply:
Good point, I was missing that context.